### PR TITLE
[Feature] Added pubsub support using watermill

### PR DESCRIPTION
**Pull request by khannaabhi**

### CODE REVIEW

File: sc-poc/modules/pubsub/app.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+// Stop ends the pubsub app operations
+func (a *App) Stop() error {
+	return nil
+}
+
+// Cleanup cleans up the app
+func (a *App) Cleanup() error {
+	return nil
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/app.go:63]: Ypu need to destroy/ remove the key from the usage pool here

File: sc-poc/modules/pubsub/connectors/connectors.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/pubsub/gochannel"
+)
+
+// Connector implements connector for pubsub module
+type Connector struct {
+	lock sync.RWMutex
+
+	// For pubsub engine
+	PubSubClient *gochannel.GoChannel
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/connectors/connectors.go:15]: Don't export this directly. Make a method to return this

File: sc-poc/modules/pubsub/operations.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+func (a *App) Subscribe(ctx context.Context, clientID, topic string, options SubscribeOptions) (<-chan *message.Message, error) {
+	messages, err := a.pubSub.Subscribe(ctx, topic)
+	if err != nil {
+		return nil, err
+	}
+	return messages, nil
+}
+
+// SendAckNack is used to send an ack/nack
+func (a *App) SendAckNack(messageID string, ack bool) error {
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/operations.go:32]: Remove this

File: sc-poc/modules/pubsub/operations.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+				Payload: ChannelPayload{
+					Schema: map[string]string{
+						"$ref": "#/components/messages/MyPayload",
+					},
+				},
+			},
+		},
+		Components: Components{
+			Messages: map[string]interface{}{
+				"type":                 "object",
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/operations.go:51]: Key should be `APIManMsg`. Its value will be the json schema object

File: sc-poc/modules/pubsub/operations.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+// Channels return channels with their schema
+func (a *App) Channels() ChannelsWithSchema {
+	return ChannelsWithSchema{
+		Channels: map[string]Channel{
+			"/sc/api": {
+				Name: "api-provision",
+				Payload: ChannelPayload{
+					Schema: map[string]string{
+						"$ref": "#/components/messages/MyPayload",
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/operations.go:44]: Rename `MyPayload` to `APIManMsg`

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: sc-poc/modules/pubsub/app.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+// Cleanup cleans up the app
+func (a *App) Cleanup() error {
+	_, err := connectorPool.Delete(getPoolKey())
+	if err != nil {
+		a.logger.Error("Unable to gracefully close  pubsub connector", zap.Error(err))
+		return err
+	}
+	return nil
+}
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/app.go:69]: Add interface guards

File: sc-poc/modules/pubsub/connectors/connectors.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+	// For pubsub engine
+	pubSubClient *gochannel.GoChannel
+}
+
+// New intialise the pubsub connector
+func New() (*Connector, error) {
+	pubSub := gochannel.NewGoChannel(
+		gochannel.Config{},
+		watermill.NewStdLogger(false, false),
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/connectors/connectors.go:22]: Is it possible to use caddy;s logger here?

File: sc-poc/modules/pubsub/app.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+	poolKey := getPoolKey()
+	val, _, err := connectorPool.LoadOrNew(poolKey, func() (caddy.Destructor, error) {
+		return connectors.New()
+	})
+	if err != nil {
+		a.logger.Error("Unable to open pubsub connector", zap.Error(err))
+		return err
+	}
+	connector := val.(*connectors.Connector)
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/app.go:45]: Set the logger here if watermill can use caddy's logger

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: sc-poc/modules/pubsub/connectors/connectors.go (hunk 7 above)
Comment (khannaabhi) [reply sc-poc/modules/pubsub/connectors/connectors.go:22]: no the methods are not same

**Pull request merged**